api/discovery/application.go

```diff
@@ -0,0 +1,103 @@
+package discovery
+
+import (
+	"context"
+
+	"github.com/dogmatiq/configkit"
+)
+
+// Application is an application configuration that is aware of the client it
+// was obtained from.
+type Application struct {
+	configkit.Application
+
+	// Client is the client that queried the application configuration.
+	Client *Client
+}
+
+// ApplicationObserver is notified when Dogma applications are discovered.
+type ApplicationObserver interface {
+	// ApplicationAvailable is called when an application becomes available.
+	ApplicationAvailable(*Application)
+
+	// ApplicationUnavailable is called when an application becomes unavailable.
+	ApplicationUnavailable(*Application)
+}
+
+// ApplicationObserverSet is an ApplicationObserver that publishes to other
+// observers.
+type ApplicationObserverSet struct {
+	observerSet
+}
+
+// NewApplicationObserverSet registers the given observers with a new observer
+// set and returns it.
+func NewApplicationObserverSet(observers ...ApplicationObserver) *ApplicationObserverSet {
+	s := &ApplicationObserverSet{}
+
+	for _, o := range observers {
+		s.RegisterApplicationObserver(o)
+	}
+
+	return s
+}
+
+// RegisterApplicationObserver registers o to be notified when applications
+// become available and unavailable.
+func (s *ApplicationObserverSet) RegisterApplicationObserver(o ApplicationObserver) {
+	s.register(o, func(e interface{}) {
+		o.ApplicationAvailable(e.(*Application))
+	})
+}
+
+// UnregisterApplicationObserver stops o from being notified when applications
+// become available and unavailable.
```

Comment:
```suggestion
// UnregisterApplicationObserver stops o from being notified when applications
// become available or unavailable.
```
Would it sound a bit better?

Reply:
I was undecided about that also. It feels like it should be "or", but then it does notify of both so I wasn't sure and just left it as is.

Reply:
Yeah I was the same, to me `and` sounds like it notifies about availability or unavailability at the same time, with or it's a bit a bit like 'either of them at a certain time'. IDK.

Reply:
I chose this word without thought, but to justify it: there are many applications involved, so some will be available and some will be unavailable, so from the perspective of the observer it's not an either/or scenario.

Reply:
Cool, clear.

Reply:
```
// become available and unavailable or why not both.
``

Reply:
But yeah that makes sense to me also.